README.md
# NAME

Router::Boom - Fast routing engine for web applications

# SYNOPSIS

    use Router::Boom;

    my $router = Router::Boom->new();
    $router->add('/', 'dispatch_root');
    $router->add('/entrylist', 'dispatch_entrylist');
    $router->add('/:user', 'dispatch_user');
    $router->add('/:user/{year}', 'dispatch_year');
    $router->add('/:user/{year}/{month:\d+}', 'dispatch_month');
    $router->add('/download/*', 'dispatch_download');

    my $dest = $router->match($env->{PATH_INFO});

# DESCRIPTION

Router::Boom is a fast path routing engine for Perl5.

# MEHTODS

- my $router = Router::Boom->new()

    Create new instance.

- $router->add($path:Str, $destination:Any)

    Add new route.

- my ($destination, $captured) = $router->match($path:Str);

    Matching the route. If matching successfully, this method returns 2 values.

    First: The destination value, you registered. Second: Captured values from the path.

    If matching was failed, this method returns empty list.

# HOW TO WRITE A ROUTING RULE

## plain string 

    $router->add( '/foo', { controller => 'Root', action => 'foo' } );

## :name notation

    $router->add( '/wiki/:page', { controller => 'WikiPage', action => 'show' } );
    ...
    $router->match('/wiki/john');
    # => {controller => 'WikiPage', action => 'show'}, {page => 'john'}

':name' notation matches `qr{([^/]+)}`.

## '\*' notation

    $router->add( '/download/*', { controller => 'Download', action => 'file' } );
    ...
    $router->match('/download/path/to/file.xml');
    # => {controller => 'Download', action => 'file'}, {'*' => 'path/to/file.xml'}

'\*' notation matches `qr{(.+)}`. You will get the captured argument as the special key: `*`.

## '{year}' notation

    $router->add( '/blog/{year}', { controller => 'Blog', action => 'yearly' } );
    ...
    $router->match('/blog/2010');
    # => {controller => 'Blog', action => 'yearly'}, {year => 2010}

'{year}' notation matches `qr{([^/]+)}`, and it will be captured.

## '{year:\[0-9\]+}' notation

    $router->add( '/blog/{year:[0-9]+}/{month:[0-9]{2}}', { controller => 'Blog', action => 'monthly' } );
    ...
    $router->match('/blog/2010/04');
    # => {controller => 'Blog', action => 'monthly'}, {year => 2010, month => '04'}

You can specify regular expressions in named captures.

Note. You can't include normal capture in custom regular expression. i.e. You can't use ` {year:(\d+)} `.
But you can use `{year:(?:\d+)}`.

# PERFORMANCE

Router::Boom is pretty fast!

                      Rate Router::Simple   Router::Boom
    Router::Simple  8000/s             --           -90%
    Router::Boom   83651/s           946%             --

Router::Boom's computational complexity is not linear scale, bug Router::Simple's computational complexity is linear scale.

Then, Router::Simple get slower if registered too much routes.
But if you're using Router::Boom then you don't care the performance :)

# LICENSE

Copyright (C) tokuhirom.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

# AUTHOR

tokuhirom <tokuhirom@gmail.com>

# SEE ALSO

[Router::Simple](https://metacpan.org/pod/Router::Simple) is my old one. But it's bit slow and complicated.

[Path::Dispatcher](https://metacpan.org/pod/Path::Dispatcher) is similar, but so complex.

[Path::Router](https://metacpan.org/pod/Path::Router) is heavy. It depends on [Moose](https://metacpan.org/pod/Moose).

[HTTP::Router](https://metacpan.org/pod/HTTP::Router) has many dependencies. It is not well documented.

[HTTPx::Dispatcher](https://metacpan.org/pod/HTTPx::Dispatcher) is my old one. It does not provide an OO-ish interface.


Changes
Revision history for Perl extension Router-Boom

{{$NEXT}}

1.03 2016-04-17T07:03:31Z

    commit 6c63b7b3f9ca1cd72f02fd3891b8a615ae58ba3a
    Author: Syohei YOSHIDA <syohex@gmail.com>
    Date:   Fri Apr 15 16:29:38 2016 +0900

        Correct return values in document

1.02 2015-06-05T07:02:54Z

    - RouterBoom#match() return allowed methods.
      (Hiroyoshi Houchi)

1.01 2014-02-24T08:07:41Z

    - Added EXPERIMENTAL 'Router::Boom::Method#routes' method.

1.00 2013-11-27T00:25:59Z

    - Bump up to 1.00.

0.07 2013-10-16T02:10:31Z

    - Router::Boom::Method accepts $router->add([qw(GET HEAD)], ...).
      (tokuhirom)

0.06 2013-10-15T23:20:45Z

    - Added Router::Boom::Method
      (tokuhirom)

0.05 2013-10-14T00:04:37Z

    - Fixed deps
      (syohex)

0.04 2013-10-12T03:08:23Z

    - no Moo; use Class::Accessor::Lite;
      (tokuhirom)

0.03 2013-10-12T01:35:35Z

    - use m// op with \G instead of s///
      (gfx)

0.02 2013-10-12T01:27:00Z

    - `{year:(?:.)}` style capturing.
      (tokuhirom)

0.01 2013-10-12T00:25:18Z

    - original version



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Router-Boom',
    module_name     => 'Router::Boom',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,


);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

use File::Copy;

print "cp META.json MYMETA.json\n";
copy("META.json","MYMETA.json") or die "Copy failed(META.json): $!";

if (-f 'META.yml') {
    print "cp META.yml MYMETA.yml\n";
    copy("META.yml","MYMETA.yml") or die "Copy failed(META.yml): $!";
} else {
    print "There is no META.yml... You may install this module from the repository...\n";
}



author/benchmark.pl
#!/usr/bin/env perl
use strict;
use warnings;
use utf8;
use 5.010000;
use autodie;
use Benchmark qw(:all);
use Router::Boom;
use Router::Boom::Method;
use Router::Simple;

my $router_boom = do {
    my $router = Router::Boom->new();
    $router->add('/',      'Root');
    $router->add('/entrylist', 'EntryList');
    $router->add("/$_", "$_") for 'a'..'z';
    $router->add('/:user', 'User#index');
    $router->add('/:user/:year/', 'UserBlog#year_archive');
    $router->add('/:user/:year/:month/', 'UserBlog#month_archive');
    $router;
};

my $router_boom_method = do {
    my $router = Router::Boom::Method->new();
    $router->add(undef, '/',      'Root');
    $router->add(undef, '/entrylist', 'EntryList');
    $router->add(undef, "/$_", "$_") for 'a'..'z';
    $router->add(undef, '/:user', 'User#index');
    $router->add(undef, '/:user/:year/', 'UserBlog#year_archive');
    $router->add(undef, '/:user/:year/:month/', 'UserBlog#month_archive');
    $router;
};

my $router_simple = do {
    my $router = Router::Simple->new();
    $router->connect('/',      'Root');
    $router->connect('/entrylist', 'EntryList');
    $router->connect("/$_", "$_") for 'a'..'z';
    $router->connect('/:user', 'User#index');
    $router->connect('/:user/:year/', 'UserBlog#year_archive');
    $router->connect('/:user/:year/:month/', 'UserBlog#month_archive');
    $router;
};

cmpthese(
    -1, {
        'Router::Simple'         => sub { $router_simple->match('/dankogai/2013/02') },
        'Router::Boom'           => sub { $router_boom->match('/dankogai/2013/02') },
        'Router::Boom::Method'   => sub { $router_boom_method->match('GET', '/dankogai/2013/02') },
    }
);



lib/Router/Boom/Method.pm
package Router::Boom::Method;
use strict;
use warnings;
use utf8;
use 5.008_005;

use Router::Boom;

our $VERSION = "1.03";

sub new {
    my $class = shift;
    my $self = bless {}, $class;
    return $self;
}

sub add {
    my ($self, $method, $path, $opaque) = @_;

    delete $self->{router}; # clear cache

    if (defined($method) && !ref($method)) {
        $method = [$method];
    }

    unless ($self->{path_seen}->{$path}++) {
        push @{$self->{path}}, $path;
    }
    push @{$self->{data}->{$path}}, [$method, $opaque];
}

sub routes {
    my $self = shift;

    my @routes;
    for my $path (@{$self->{path}}) {
        for my $route (@{$self->{data}->{$path}}) {
            my ($method, $opaque) = @$route;
            push @routes, [ $method, $path, $opaque ];
        }
    }
    return @routes;
}

sub _method_match {
    my ($request_method, $matcher) = @_;
    return 1 if not defined($matcher);

    for my $m (@$matcher) {
        return 1 if $m eq $request_method;
    }
    return 0;
}

sub match {
    my ($self, $request_method, $path) = @_;

    $self->{router} ||= $self->_build_router();

    if (my ($patterns, $captured) = $self->{router}->match($path)) {
        my @allowed_methods;
        for my $pattern (@$patterns) {
            if (_method_match($request_method, $pattern->[0])) {
                return ($pattern->[1], $captured, 0, []);
            }
            push @allowed_methods, @{$pattern->[0]};
        }
        return (undef, undef, 1, \@allowed_methods);
    } else {
        return;
    }
}

sub regexp {
    my $self = shift;
    $self->{router} ||= $self->_build_router();
    $self->{router}->regexp;
}

sub _build_router {
    my ($self) = @_;
    my $router = Router::Boom->new();
    for my $path (@{$self->{path}}) {
        $router->add($path, $self->{data}->{$path});
    }
    $router;
}

1;
__END__

=head1 NAME

Router::Boom::Method - Router::Boom with HTTP method support

=head1 DESCRIPTION

Router::Boom doesn't care the routing with HTTP method. It's simple and good.
But it makes hard to implement the rule like this:

    get  '/' => sub { 'get ok'  };
    post '/' => sub { 'post ok' };

Then, this class helps you.

=head1 METHODS

=over 4

=item C<< my $router = Router::Boom::Method->new() >>

Create new instance.

=item C<< $router->add($http_method:Str|ArrayRef[Str], $path:Str, $opaque:Any) >>

Add new path to the router.

C<$http_method> is a string to represent HTTP method. i.e. GET, POST, DELETE, PUT, etc.
The path can handle any HTTP methods, you'll path the C<undef> for this argument.
You can specify the multiple HTTP methods in ArrayRef like C<< $router->add([qw(GET HEAD)], '/', 'top') >>.
It will be matching with the C<REQUEST_METHOD>.

C<$path> is the path string. It will be matching with the C<PATH_INFO>.

C<$opaque> is the destination path data. Any data is OK.

=item C<< my ($dest, $captured, $is_method_not_allowed, $allowed_methods) = $router->match($http_method:Str, $path:Str) >>

Matching with the router.

C<$http_method> is the HTTP request method. It's C<< $env->{REQUEST_METHOD} >> in PSGI.

C<$path> is the path string. It's C<< $env->{PATH_INFO} >> in PSGI.

I<Return Value:>

If the request is not matching with any path, this method returns empty list.

If the request is matched well then, return C<$dest>, C<$captured>. And C<$is_method_not_allowed> is false value.

If the request path is matched but the C<$http_method> is not matched, then C<$dest> and C<$captured> is undef. And C<$is_method_not_allowed> is true value. You got this then you need to return C<405 Method Not Allowed> error.

If the request path is matched but the C<$http_method> is not matched, then C<$dest> and C<$captured> is undef. And C<$allowed_methods> is ArrayRef. You got this then you need to return C<405 Method Not Allowed> error with C<Allow> header.

=item C<< my $regexp = $router->regexp() >>

Get a compiled regexp for debugging.

=item C<< my @routes = $router->routes() >>

B<EXPERIMENTAL>

Get the list of registered routes. Every routes has following schema.

    [Maybe[ArrayRef], Str, Any]

For example:

    [['GET','HEAD'], "/foo", \&dispatch_foo]

=back

=head1 AUTHORS

Tokuhiro Matsuno



lib/Router/Boom.pm
package Router::Boom;
use 5.008005;
use strict;
use warnings;
use Carp ();

our $VERSION = "1.03";

# Matcher stuff
our $LEAF_IDX;
our @CAPTURED;

# Compiler stuff
our @LEAVES;
our $PAREN_CNT;
our @PARENS;

use re 'eval';

use Router::Boom::Node;

sub new {
    my $class = shift;
    my $self = bless { }, $class;
    $self->{root} = Router::Boom::Node->new(key => '/');
    return $self;
}

# True if : ()
# False if : (?:)
sub _is_normal_capture {
    $_[0] =~ /
        \(
            (?!
                \?:
            )
    /x
}

sub add {
    my ($self, $path, $stuff) = @_;
    $path =~ s!\A/!!;

    delete $self->{regexp}; # clear cache

    my $p = $self->{root};
    my @capture;
    while ($path =~ m!\G(?:
            \{((?:\{[0-9,]+\}|[^{}]+)+)\} | # /blog/{year:\d{4}}
            :([A-Za-z0-9_]+)              | # /blog/:year
            (\*)                          | # /blog/*/*
            ([^{:*]+)                       # normal string
        )!xg) {

        if (defined $1) {
            my ($name, $pattern) = split /:/, $1, 2;
            if (defined($pattern) && _is_normal_capture($pattern)) {
                Carp::croak("You can't include parens in your custom rule.");
            }
            push @capture, $name;
            $pattern = $pattern ? "($pattern)" : "([^/]+)";
            $p = $p->add_node($pattern);
        } elsif (defined $2) {
            push @capture, $2;
            $p = $p->add_node("([^/]+)");
        } elsif (defined $3) {
            push @capture, '*';
            $p = $p->add_node("(.+)");
        } else {
            $p = $p->add_node(quotemeta $4);
        }
    }
    $p->leaf([\@capture, $stuff]);

    return;
}

sub _build_regexp {
    my ($self) = @_;

    my $trie = $self->{root};
    local @LEAVES;
    local $PAREN_CNT = 0;
    local @PARENS;
    my $re = _to_regexp($trie);
    $self->{leaves} = [@LEAVES];
    $self->{regexp} = qr{\A$re};
}

sub match {
    my ($self, $path) = @_;

    # "I think there was a discussion about that a while ago and it is up to apps to deal with empty PATH_INFO as root / iirc"
    # -- by @miyagawa
    #
    # see http://blog.64p.org/entry/2012/10/05/132354
    $path = '/' if $path eq '';

    if ($path =~ $self->regexp) {
        my ($captured, $stuff) = @{$self->{leaves}->[$Router::Boom::LEAF_IDX]};
        my %captured;
        @captured{@$captured} = @Router::Boom::CAPTURED;
        return ($stuff, \%captured);
    } else {
        return ();
    }
}

sub regexp {
    my $self = shift;
    if (not exists $self->{regexp}) {
        $self->_build_regexp();
    }
    $self->{regexp};
}

sub _to_regexp {
    my ($node) = @_;

    my %leaves;

    local @PARENS = @PARENS;

    my $key = $node->key;
    if ($key =~ /\(/) {
        $PAREN_CNT++;
        push @PARENS, $PAREN_CNT;
    }
    my @re;
    if (@{$node->children}>0) {
        push @re, map { _to_regexp($_) } @{$node->children};
    }
    if ($node->leaf) {
        push @Router::Boom::LEAVES, $node->leaf;
        push @re, sprintf '\z(?{ $Router::Boom::LEAF_IDX=%s; @Router::Boom::CAPTURED = (%s) })', @Router::Boom::LEAVES-1, join(',', map { "\$$_" } @PARENS);
    }
    my $re = $node->key;
    if (@re==0) {
        # nop
    } elsif (@re == 1) {
        $re .= $re[0];
    } else {
        $re .= '(?:' . join('|', @re) . ')';
    }
    return qr{$re};
}

1;
__END__

=encoding utf-8

=head1 NAME

Router::Boom - Fast routing engine for web applications

=head1 SYNOPSIS

    use Router::Boom;

    my $router = Router::Boom->new();
    $router->add('/', 'dispatch_root');
    $router->add('/entrylist', 'dispatch_entrylist');
    $router->add('/:user', 'dispatch_user');
    $router->add('/:user/{year}', 'dispatch_year');
    $router->add('/:user/{year}/{month:\d+}', 'dispatch_month');
    $router->add('/download/*', 'dispatch_download');

    my $dest = $router->match($env->{PATH_INFO});

=head1 DESCRIPTION

Router::Boom is a fast path routing engine for Perl5.

=head1 MEHTODS

=over 4

=item my $router = Router::Boom->new()

Create new instance.

=item $router->add($path:Str, $destination:Any)

Add new route.

=item my ($destination, $captured) = $router->match($path:Str);

Matching the route. If matching successfully, this method returns 2 values.

First: The destination value, you registered. Second: Captured values from the path.

If matching was failed, this method returns empty list.

=back

=head1 HOW TO WRITE A ROUTING RULE

=head2 plain string 

    $router->add( '/foo', { controller => 'Root', action => 'foo' } );

=head2 :name notation

    $router->add( '/wiki/:page', { controller => 'WikiPage', action => 'show' } );
    ...
    $router->match('/wiki/john');
    # => {controller => 'WikiPage', action => 'show'}, {page => 'john'}

':name' notation matches C<qr{([^/]+)}>.

=head2 '*' notation

    $router->add( '/download/*', { controller => 'Download', action => 'file' } );
    ...
    $router->match('/download/path/to/file.xml');
    # => {controller => 'Download', action => 'file'}, {'*' => 'path/to/file.xml'}

'*' notation matches C<qr{(.+)}>. You will get the captured argument as the special key: C<*>.

=head2 '{year}' notation

    $router->add( '/blog/{year}', { controller => 'Blog', action => 'yearly' } );
    ...
    $router->match('/blog/2010');
    # => {controller => 'Blog', action => 'yearly'}, {year => 2010}

'{year}' notation matches C<qr{([^/]+)}>, and it will be captured.

=head2 '{year:[0-9]+}' notation

    $router->add( '/blog/{year:[0-9]+}/{month:[0-9]{2}}', { controller => 'Blog', action => 'monthly' } );
    ...
    $router->match('/blog/2010/04');
    # => {controller => 'Blog', action => 'monthly'}, {year => 2010, month => '04'}

You can specify regular expressions in named captures.

Note. You can't include normal capture in custom regular expression. i.e. You can't use C< {year:(\d+)} >.
But you can use C<< {year:(?:\d+)} >>.

=head1 PERFORMANCE

Router::Boom is pretty fast!

                      Rate Router::Simple   Router::Boom
    Router::Simple  8000/s             --           -90%
    Router::Boom   83651/s           946%             --

Router::Boom's computational complexity is not linear scale, bug Router::Simple's computational complexity is linear scale.

Then, Router::Simple get slower if registered too much routes.
But if you're using Router::Boom then you don't care the performance :)

=head1 LICENSE

Copyright (C) tokuhirom.

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=head1 AUTHOR

tokuhirom E<lt>tokuhirom@gmail.comE<gt>

=head1 SEE ALSO

L<Router::Simple> is my old one. But it's bit slow and complicated.

L<Path::Dispatcher> is similar, but so complex.

L<Path::Router> is heavy. It depends on L<Moose>.

L<HTTP::Router> has many dependencies. It is not well documented.

L<HTTPx::Dispatcher> is my old one. It does not provide an OO-ish interface.

=cut

